### PR TITLE
Close Qt progress bars when viewer is closed

### PR DESCRIPTION
**Pull request by DragaDoncila**
# Description
As described in #2884 , closing the viewer while a progress bar is active currently results in a RuntimeError due to the `progress` object persisting after the widgets are all closed. 

When iteration continues and the various update methods are called, the `ProgressBar` widget has already been deleted.

This PR
- removes the `self.has_viewer` attribute as this attribute really stored whether or not we had a `ProgressBar` widget associated with this `progress` object. We instead check `self._pbar` directly.
- adds `close_pbar` method separate to `close` method of `progress` so that we can close the `ProgressBar` widget without necessarily closing the `progress` object. This method also sets the `pbar` to `None`.
- connects `close_pbar` to `viewer.destroyed` when the `ProgressBar` widget is created, rather than closing the progress object entirely
- wraps `display` method code which accesses the `ProgressBar` widget's attributes in a `try/except` block. Checking that `self._pbar is not None` before accessing these attributes is insufficient because of the non-sequential code path taken by `tqdm` 

This is a very bandaidy fix and I think longer term we may wish to consider no longer inheriting from `tqdm`, or overriding a good chunk of the `update/refresh` methods. While it provides a lot of convenience, it also takes a good chunk of code out of our explicit control. 

@tlambert03 I'm curious what you think about this fix, the necessity of the `try/except` and how we should refactor the `progress` structure to be less brittle and more tightly coupled with the viewer. Maybe we could pair at some stage and discuss.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by DragaDoncila** [comment]
@sofroniewn do we want to review and potentially merge this while we further discuss #3004 and whether we actually want to go through with it?

**Review by sofroniewn — APPROVED** [review]
LGTM! Agreed, I think we should get this in now and then we can decide about tqdm separately. I'll wait for @tlambert03 to look at this before merge, but I think this is good to go in, all looked good to me!!

**Review by tlambert03 — APPROVED** [review]
Thanks @DragaDoncila!  I think this steps in a good direction

> @tlambert03 I'm curious what you think about this fix, the necessity of the try/except and how we should refactor the progress structure to be less brittle and more tightly coupled with the viewer. Maybe we could pair at some stage and discuss.

I remember playing with that try/except specifically in `display`, and was unable to come up with a cleaner method, so I think that's where we should start for now.  maybe if we drop tqdm it will be easier, maybe not.  but this is better already.

i feel like we should keep working on this until  you can open up the pbar example, quit in the middle of a progress, and have it not throw errors.  But at least now it quits! :)

**Pull request merged**
